README
NAME
    Games::Dice::Tester - Statistical tests of random-dice-rolling programs

    Various statistical tests for use with dice-rolling programs.

WHAT THIS MODULE IS NOT
    A way to tell you whether your random-number generator is truly random!

EXPORT
    Nothing by default; ":all" for everything; otherwise, any of the functions
    listed below.

FUNCTIONS
  chi_squared
    Compute the chi-squared statistic V.  Usage:

        my ($chi_squared, $df) = chi_squared_test(observed=>[...], expected=>[...]);

    where the "observed" and "expected" arrays are the number of occurrences
    in each category.

    As a shorthand for the common case of a fair die, you can pass a scalar
    for "expected".  The "expected" array will then be filled with copies of
    that value equal in number to the size of the "observed" array.

    As yet another alternative, you can call

        my ($chi_squared, $df) = chi_squared_test(generator=>sub { ... });

    where "generator", when called, returns the next "($observed, $expected)"
    pair.  When the generator returns "undef", the loop stops.

    Returns the chi-squared statistic.  In list context, also returns the
    number of degrees of freedom.

    Original version by  Lukas Atkinson
    <https://stackoverflow.com/users/1521179/amon>, posted here
    <https://stackoverflow.com/a/21205042/2877364>.  Modified by CXW.

  chi_squared_test
    Return the probability that a given observed distribution was produced by
    a random process with the given expected distribution.  Parameters are as
    "chi_squared".

    If I am reading Knuth correctly, a random sequence should generally have
    probabilities on the range (0.1, 0.9).  TAOCP 3e, vol. 2, p. 47.

    Original version by  Lukas Atkinson
    <https://stackoverflow.com/users/1521179/amon>, posted here
    <https://stackoverflow.com/a/21205042/2877364>.  Modified by CXW.

  chi_squared_maybe_ok
    Parameters as "chi_squared_test".  Returns true iff the chi-squared result
    is in the range (0.1, 0.9).

SEE ALSO
    *   <http://blogs.perl.org/users/ovid/2014/01/testing-random-dice-rolls.ht
        ml>

    *   <https://www.nu42.com/2014/01/randomness-and-statistical-concepts.html
        >

    *   <http://oddgeek.info/code/gdp/>

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/cxw42/Games-Dice-Tester/issues>.  I will be notified,
    and then you'll automatically be notified of progress on your bug as I
    make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Games::Dice::Tester

    You can also look for information at:

    *   GitHub (main repository)

        <https://github.com/cxw42/Games-Dice-Tester>

    *   MetaCPAN

        <https://metacpan.org/release/Games-Dice-Tester>

    *   CPAN Ratings

        <https://cpanratings.perl.org/d/Games-Dice-Tester>

LICENSE
    Copyright 2019 Christopher White.  Some code Copyright 2014 Lukas
    Atkinson.

    Licensed CC-BY-SA 3.0 <https://creativecommons.org/licenses/by-sa/3.0/>.

    Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
    AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. 
    THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
    PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
    YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR
    WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR CONSEQUENTIAL
    DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE, EVEN IF ADVISED
    OF THE POSSIBILITY OF SUCH DAMAGE.

POD ERRORS
    Hey! The above document had some coding errors, which are explained below:

    Around line 72:
        L<> starts or ends with whitespace

    Around line 124:
        L<> starts or ends with whitespace



Changes
Revision history for Games-Dice-Tester

0.000001    2019-??-??
        First version, released on an unsuspecting world.



t/MY/Kit.pm
#!perl
# Kit: Test kit for Games::Dice::Tester
package MY::Kit;

use 5.014;
use strict;
use warnings;
#use Test::Exception ();
#use Test::Fatal ();
use Test::More ();
use Test::Number::Delta ();

use Import::Into;

sub import {
    my $target = caller;
    $_->import::into($target) foreach
        qw(strict warnings Test::More Test::Number::Delta);
            # Test::Exception Test::Fatal
} #import()

1;


lib/Games/Dice/Tester.pm
package Games::Dice::Tester;

use 5.014;
use strict;
use warnings;
use parent 'Exporter';

our $VERSION = '0.000001';  # TRIAL

our (@EXPORT, @EXPORT_OK, %EXPORT_TAGS);
BEGIN {
    @EXPORT = ();
    @EXPORT_OK = qw(chi_squared chi_squared_test chi_squared_maybe_ok);
    %EXPORT_TAGS = (
        default => [@EXPORT],
        all => [@EXPORT, @EXPORT_OK]
    );
}

use Carp qw< croak >;
use List::Util qw< sum >;
use Statistics::Distributions qw< chisqrprob >;

# Docs =================================================================== {{{1

=head1 NAME

Games::Dice::Tester - Statistical tests of random-dice-rolling programs

=head1 SYNOPSIS

Various statistical tests for use with dice-rolling programs.

=head1 WHAT THIS MODULE IS NOT

A way to tell you whether your random-number generator is truly random!

=head1 EXPORT

Nothing by default; C<:all> for everything; otherwise, any of the functions
listed below.

=head1 FUNCTIONS

=cut

# }}}1

=head2 chi_squared

Compute the chi-squared statistic V.  Usage:

    my ($chi_squared, $df) = chi_squared_test(observed=>[...], expected=>[...]);

where the C<observed> and C<expected> arrays are the number of occurrences
in each category.

As a shorthand for the common case of a fair die, you can pass a scalar for
C<expected>.  The C<expected> array will then be filled with copies of that
value equal in number to the size of the C<observed> array.

As yet another alternative, you can call

    my ($chi_squared, $df) = chi_squared_test(generator=>sub { ... });

where C<generator>, when called, returns the next C<($observed, $expected)>
pair.  When the generator returns C<undef>, the loop stops.

Returns the chi-squared statistic.  In list context, also returns
the number of degrees of freedom.

Original version by L< Lukas
Atkinson|https://stackoverflow.com/users/1521179/amon>, posted
L<here|https://stackoverflow.com/a/21205042/2877364>.  Modified by CXW.

=cut

sub chi_squared {
    my %args = @_;
    my ($observed, $expected, $generator);

    if($args{generator}) {
        croak "Generator must be a coderef" unless ref $args{generator} eq 'CODE';
        $generator = delete $args{generator};
        croak 'Generator must be the only parameter' if scalar keys %args;

    } else {
        my $idx = 0;
        $observed = delete $args{observed} // croak q(Argument "observed" required);
        $expected = delete $args{expected} // croak q(Argument "expected" required);
        $generator = sub { ++$idx; ($observed->[$idx-1], $expected->[$idx-1]) };

        # Shorthand syntax for $expected
        $expected = [($expected) x @$observed] unless ref $expected eq 'ARRAY';

        # Validate parameters
        @$observed == @$expected or croak q(Input arrays must have same length);
    }

    # Do the work
    my $count = 0;
    my $chi_squared;
    while(1) {
        my ($obs, $exp) = $generator->();
        last unless defined $obs;
        ++$count;
        $chi_squared += ($obs - $exp)**2 / $exp;
    }

    my $degrees_of_freedom = $count - 1;

    return wantarray ? ($chi_squared, $degrees_of_freedom) : $chi_squared;
} #chi_squared()

=head2 chi_squared_test

Return the probability that a given observed distribution was produced by a
random process with the given expected distribution.  Parameters are as
L</chi_squared>.

If I am reading Knuth correctly, a random sequence should generally have
probabilities on the range (0.1, 0.9).  TAOCP 3e, vol. 2, p. 47.

Original version by L< Lukas
Atkinson|https://stackoverflow.com/users/1521179/amon>, posted
L<here|https://stackoverflow.com/a/21205042/2877364>.  Modified by CXW.

=cut

sub chi_squared_test {
    my ($chi_squared, $degrees_of_freedom) = chi_squared(@_);
    my $probability = chisqrprob($degrees_of_freedom, $chi_squared);
    return $probability;
} #chi_squared_test()

=head2 chi_squared_maybe_ok

Parameters as L</chi_squared_test>.  Returns true iff the chi-squared result
is in the range (0.1, 0.9).

=cut

sub chi_squared_maybe_ok {
    my $prob = chi_squared_test(@_);
    return ($prob > 0.1) && ($prob < 0.9);
}

1; # End of Games::Dice::Tester
__END__

# Rest of docs =========================================================== {{{1

=head1 SEE ALSO

=over

=item *

L<http://blogs.perl.org/users/ovid/2014/01/testing-random-dice-rolls.html>

=item *

L<https://www.nu42.com/2014/01/randomness-and-statistical-concepts.html>

=item *

L<http://oddgeek.info/code/gdp/>

=back

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/cxw42/Games-Dice-Tester/issues>.  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Dice::Tester

You can also look for information at:

=over 4

=item * GitHub (main repository)

L<https://github.com/cxw42/Games-Dice-Tester>

=item * MetaCPAN

L<https://metacpan.org/release/Games-Dice-Tester>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Games-Dice-Tester>

=back

=head1 LICENSE

Copyright 2019 Christopher White.
Some code Copyright 2014 Lukas Atkinson.

Licensed L<CC-BY-SA 3.0|https://creativecommons.org/licenses/by-sa/3.0/>.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

# }}}1
# vi: set fdm=marker: #


Makefile.PL
# Makefile.PL for Games::Dice::Tester
use 5.014;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use File::Spec;
use Config;

my $VERSION_FROM = File::Spec->catfile(qw(lib Games Dice Tester.pm));
my $IS_TRIAL = check_trial();
my $secure_perl_path = get_perl_filename();
my @provides = ();

# Check if this is a TRIAL version {{{1
sub check_trial {
    TRIAL: {
        # Get the VERSION line
        open my $fd, '<', $VERSION_FROM or last TRIAL;
        my $linetext;
        while(<$fd>) {
            next unless /VERSION/;
            $linetext = $_;
            last;
        }
        close $fd;
        return !!($linetext =~ /\bTRIAL\b/);
    }
    return 0;
} #check_trial()

# }}}1
# Module metadata {{{1
eval {
    require Module::Metadata;
    @provides =
        (provides => Module::Metadata->provides(version => '2', dir => 'lib'));
        # Thanks to https://stackoverflow.com/a/28928985/2877364 by LEONT
        # for suggesting Module::Metadata.
};

# }}}1
# Get the filename of the Perl interpreter running this. {{{1
# Modified from perlvar.
# The -x test is for cygwin or other systems where $Config{perlpath} has no
# extension and $Config{_exe} is nonempty.  E.g., symlink perl->perl5.10.1.exe.
# There is no "perl.exe" on such a system.
sub get_perl_filename {
    my $secure_perl_path = $Config{perlpath};
    if ($^O ne 'VMS') {
        $secure_perl_path .= $Config{_exe}
            unless (-x $secure_perl_path) ||
                            ($secure_perl_path =~ m/$Config{_exe}$/i);
    }
    return $secure_perl_path;
} # get_perl_filename()

# }}}1
# Makefile customization (MY) {{{1
{   package MY;

    # dist_core: make `dist` a :: target rather than a : target,
    # and add distcheck before dist.  Also, add -TRIAL to the tgz if necessary.
    sub dist_core {
        my $self = shift;
        my $text = $self->SUPER::dist_core(@_);
        $text =~ s/^dist\s*:[^:]/dist:: distcheck /m;

        # Add -TRIAL if it's a trial release
        if($IS_TRIAL) {
            # Note: we don't have to worry about EOL; Appveyor uses gmake even
            # on Windows, and it only uses \n.
            print STDERR "TRIAL version\n";
            my $newtext .=  # Command to rename the tgz.  TODO see if this works on Windows.
                "\t\"$secure_perl_path\" -MExtUtils::Command -e mv -- " .
                '"$(DISTVNAME).tar$(SUFFIX)" ' .
                '"$(DISTVNAME)-TRIAL.tar$(SUFFIX)"' . "\n";

            # Insert $newtext at the end of the `dist` target
            $text =~ s{
                ^(dist\h*:.*\n      # dist header line
                    (?:\h+\S.*\n)+) # dist body lines.  `.` doesn't match `\n`.
                                    # NOTE: on Appveyor, the continuation line
                                    # begins with a space rather than a tab.
                                    # Therefore, look for \h after \n.
                                    # Not \s, because that matches a \n!
            }{$1$newtext}mx;
        } #endif $IS_TRIAL

        return $text;
    } # MY::dist_core

    # Generate README.md; add uninstall target and other test targets
    sub postamble {     # TODO also handle Windows nmake syntax (SET vs. export)

        # Note: pod2markdown is marked with a hyphen so that `make` can
        # succeed and generate META files even if Pod::Markdown is not yet
        # installed.

        my $make_readme_md = File::Spec->catfile(qw(support readme.pl));
        return <<EOT;
authortest:
\tRELEASE_TESTING=1 prove -lvj4 xt

testhere:   # Run the tests from lib rather than blib
\tprove -lj4

README.md: @{[$VERSION_FROM]} Makefile.PL $make_readme_md
\t"$secure_perl_path" "$make_readme_md" -i "\$<" -o "\$@" -f md

README: @{[$VERSION_FROM]} Makefile.PL $make_readme_md
\t"$secure_perl_path" "$make_readme_md" -i "\$<" -o "\$@" -f text

all :: README.md README
dist :: README.md README

# Uninstallation: mimic what the normal "uninstall" target does, but actually
# carry out the actions.
# Note: this may leave empty directories in place, since the packlist only
# includes files.

ACTUALLY_UNINSTALL = "$secure_perl_path" -MExtUtils::Install -e 'uninstall(shift,1)' --

actually_uninstall :: actually_uninstall_from_\$(INSTALLDIRS)dirs
\t\$(NOECHO) \$(NOOP)

actually_uninstall_from_perldirs ::
\t-\$(ACTUALLY_UNINSTALL) "\$(PERL_ARCHLIB)/auto/\$(FULLEXT)/.packlist"

actually_uninstall_from_sitedirs ::
\t-\$(ACTUALLY_UNINSTALL) "\$(SITEARCHEXP)/auto/\$(FULLEXT)/.packlist"

actually_uninstall_from_vendordirs ::
\t-\$(ACTUALLY_UNINSTALL) "\$(VENDORARCHEXP)/auto/\$(FULLEXT)/.packlist"

EOT

    } #postamble()
} #package MY

# }}}1

# Main options for EUMM
my %opts = (
    NAME             => 'Games::Dice::Tester',
    AUTHOR           => q{Christopher White <cxwembedded@gmail.com>},
    VERSION_FROM     => $VERSION_FROM,
    ABSTRACT_FROM    => $VERSION_FROM,
    LICENSE          => 'open_source',
    PL_FILES         => {},

    MIN_PERL_VERSION => '5.014',
    CONFIGURE_REQUIRES => {
        'Config' => '0',
        'ExtUtils::MakeMaker' => '0',
        'strict' => '0',
        'warnings' => '0',
    },
    #BUILD_REQUIRES => {
    #},
    TEST_REQUIRES => {
        'Games::Dice' => '0.041',
        'Games::Dice::Advanced' => '1.1',
        #'Games::Dice::Probability' => '0',
        'lib::relative' => '0.002',
        'Test::More' => '0',
        'Test::Number::Delta' => '1.06',
    },
    PREREQ_PM => {
        'Carp' => '0',
        'Exporter' => '0',
        'List::Util' => '1.50',
        'parent' => '0',
        #'Statistics::Descriptive' => '3.0700',
        'Statistics::Distributions' => '0.07',
        #'Statistics::Test::RandomWalk' => '0.02',
        #'Statistics::Test::Sequence' => '0.01',
        #'Test::Probability' => '0.002',     # part of Games::Dice::Loaded
    },

    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Games-Dice-Tester-*' },

    META_MERGE => {
        'meta-spec' => { version => 2 },

        resources => {
            bugtracker => {
                web  => 'https://github.com/cxw42/Games-Dice-Tester/issues',
            },
            repository => {
                type => 'git',
                url  => 'https://github.com/cxw42/Games-Dice-Tester.git',
                web  => 'https://github.com/cxw42/Games-Dice-Tester',
            },
        },

        prereqs => {
            develop => {
                requires => {
                    'App::RewriteVersion' => '0',   # for perl-bump-version
                    'Module::Metadata' => '1.000016',
                },
            },
        },

        @provides,

    }, #META_MERGE
);

$opts{META_MERGE}->{release_status} = 'testing' if $IS_TRIAL;

WriteMakefile(%opts);
# vi: set fdm=marker fdl=0: #
